adal/src/telemetry/java/com/microsoft/aad/adal/APIEvent.java

```diff
@@ -0,0 +1,172 @@
+            setAuthorityType(EventStrings.AUTHORITY_TYPE_AAD);
+        }
+    }
+
+    void setAuthorityType(final String authorityType) {
+        setEvent(EventStrings.AUTHORITY_TYPE, authorityType);
+    }
+
+    void setIsDeprecated(final Boolean isDeprecated) {
+        setEvent(EventStrings.API_DEPRECATED, isDeprecated.toString());
+    }
+
+    void setValidationStatus(final String validationStatus) {
+        setEvent(EventStrings.AUTHORITY_VALIDATION, validationStatus);
+    }
+
+    void setExtendedExpiresOnSetting(final Boolean extendedExpiresOnSetting) {
+        setEvent(EventStrings.EXTENDED_EXPIRES_ON_SETTING, extendedExpiresOnSetting.toString());
+    }
+
+    void setSilentRequestPromptBehavior(final String promptBehavior) {
+        setEvent(EventStrings.PROMPT_BEHAVIOR, promptBehavior);
+    }
+
+    void setAPIId(final String id) {
+        setEvent(EventStrings.API_ID, id);
+    }
+
+    String getEventName() {
+        return mEventName;
+    }
+
+    void setWasApiCallSuccessful(final Boolean isSuccess, final Exception exception) {
+        setEvent(EventStrings.WAS_SUCCESSFUL, isSuccess.toString());
+
+        if (exception != null) {
+            if (exception instanceof AuthenticationException) {
+                final AuthenticationException authException = (AuthenticationException) exception;
+                setEvent(EventStrings.API_ERROR_CODE, authException.getCode().toString());
+            }
+        }
+    }
+
+    void stopTelemetryAndFlush() {
+        Telemetry.getInstance().stopEvent(getRequestId(), this, getEventName());
+        Telemetry.getInstance().flush(getRequestId());
+    }
+
+    void setIdToken(final String rawIdToken) {
+        if (StringExtensions.isNullOrBlank(rawIdToken)) {
+            return;
+        }
+
+        final IdToken idToken;
+        try {
+            idToken = new IdToken(rawIdToken);
+        } catch (AuthenticationException ae) {
+            return;
+        }
+
+        final UserInfo userInfo = new UserInfo(idToken);
+        setEvent(EventStrings.IDP_NAME, idToken.getIdentityProvider());
+
+        try {
+            setEvent(EventStrings.TENANT_ID, StringExtensions.createHash(idToken.getTenantId()));
+            setEvent(EventStrings.USER_ID, StringExtensions.createHash(userInfo.getDisplayableId()));
+        } catch (UnsupportedEncodingException | NoSuchAlgorithmException e) {
+            Logger.i(TAG, "Skipping TENANT_ID and USER_ID", "");
+        }
+    }
+
+    void setLoginHint(final String loginHint) {
+        try {
+            setEvent(EventStrings.LOGIN_HINT,  StringExtensions.createHash(loginHint));
+        } catch (UnsupportedEncodingException | NoSuchAlgorithmException e) {
+            Logger.i(TAG, "Skipping telemetry for LOGIN_HINT", "");
+        }
+    }
+
+    @Override
+    public void processEvent(final Map<String, String> dispatchMap) {
+        super.processEvent(dispatchMap);
+        final List eventList = getEventList();
```

Comment:
We can avoid the cast at L154 by specifying generic types to this `List` with `List<Pair<String, String>>`